### PR TITLE
[core] [2/N] Skip GCS health check if possible

### PR DESCRIPTION
**Pull request by dentiny**
Followup PR for https://github.com/ray-project/ray/pull/49122
Resolves issue: https://github.com/ray-project/ray/issues/48837

This PR integrates grpc completion callback for ray syncer with health check manager, which is used to skip a few unnecessary rpcs.

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_health_check_manager.cc [hunk 1]
```diff
@@ -82,15 +82,39 @@ std::vector<NodeID> GcsHealthCheckManager::GetAllNodes() const {
   return nodes;
 }
 
+void GcsHealthCheckManager::MarkNodeHealthy(const NodeID &node_id) {
```
Comment (rynewang) [inline src/ray/gcs/gcs_server/gcs_health_check_manager.cc:85]: this method is called from another thread (ray syncer's thread). do we need a mutex?

Comment (dentiny) [reply src/ray/gcs/gcs_server/gcs_health_check_manager.cc:85]: The call comes from io context passed down to syner: https://github.com/ray-project/ray/blob/8ab5b2b38348734c90385cac72671861b5b1d67e/src/ray/common/ray_syncer/ray_syncer_bidi_reactor_base.h#L190-L192

Aren't  health check manager and syncer shared the same io context?

Comment (dentiny) [reply src/ray/gcs/gcs_server/gcs_health_check_manager.cc:85]: OK, I found they're actually different io contexts:
- ray syncer: https://github.com/ray-project/ray/blob/8ab5b2b38348734c90385cac72671861b5b1d67e/src/ray/gcs/gcs_server/gcs_server.cc#L524-L525
- health check manager: https://github.com/ray-project/ray/blob/8ab5b2b38348734c90385cac72671861b5b1d67e/src/ray/gcs/gcs_server/gcs_server.cc#L291-L292

Comment (dentiny) [reply src/ray/gcs/gcs_server/gcs_health_check_manager.cc:85]: Updated, I posted it to io context, which is the same implementation as `AddNode`.
https://github.com/ray-project/ray/blob/8ab5b2b38348734c90385cac72671861b5b1d67e/src/ray/gcs/gcs_server/gcs_health_check_manager.cc#L143-L153

File: src/ray/gcs/gcs_server/gcs_health_check_manager.cc [hunk 2]
```diff
@@ -82,15 +82,45 @@ std::vector<NodeID> GcsHealthCheckManager::GetAllNodes() const {
...
+  io_service_.dispatch(
+      [this, node_id]() {
+        thread_checker_.IsOnSameThread();
```
Comment (rynewang) [inline src/ray/gcs/gcs_server/gcs_health_check_manager.cc:88]: need RAY_CHECK ?

File: src/ray/gcs/gcs_server/gcs_health_check_manager.cc [hunk 3]
```diff
@@ -82,15 +82,45 @@ std::vector<NodeID> GcsHealthCheckManager::GetAllNodes() const {
...
+        }
+
+        auto *ctx = iter->second;
+        ctx->SetLatestHealthTimestamp(absl::Now());
+      },
+      "GcsHealthCheckManager::MarkNodeHealthy");
+}
+
 void GcsHealthCheckManager::HealthCheckContext::StartHealthCheck() {
   using ::grpc::health::v1::HealthCheckResponse;
```
Comment (rynewang) [inline src/ray/gcs/gcs_server/gcs_health_check_manager.cc:106]: also use thread_checker_ here?

File: src/ray/gcs/gcs_server/gcs_health_check_manager.cc (hunk 2 above)
Comment (rynewang) [reply src/ray/gcs/gcs_server/gcs_health_check_manager.cc:88]: can you also add [[nodiscard]] to the checker method definition

Comment (dentiny) [reply src/ray/gcs/gcs_server/gcs_health_check_manager.cc:88]: makes sense! fixed.

**Comment by rynewang** [comment]
build failure

**Comment by dentiny** [comment]
> build failure

Nice, it means we do have unchecked assertion. Should be fixed.

File: src/ray/gcs/gcs_server/gcs_health_check_manager.h [hunk 4]
```diff
@@ -82,6 +82,12 @@ class GcsHealthCheckManager {
   /// \return A list of node id which are being monitored by this class.
   std::vector<NodeID> GetAllNodes() const;
 
+  /// Mark the given node as healthy, so health check manager could save some checking
+  /// rpcs. Notice: have to invoke from io-context thread.
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_health_check_manager.h:86]: I don't think this comment is accurate

**Review by jjyao — COMMENTED** [review]
LG

File: src/ray/gcs/gcs_server/gcs_health_check_manager.h (hunk 4 above)
Comment (dentiny) [reply src/ray/gcs/gcs_server/gcs_health_check_manager.h:86]: I think it's describing what it does? Any detailed comment on which part inaccurate?

Comment (dentiny) [reply src/ray/gcs/gcs_server/gcs_health_check_manager.h]: Fixed.

**Review by jjyao — APPROVED** [review]

**Pull request merged**